### DISCUSSION TITLE
splitAtom - Inside a writable atom after fetching data from server, how to create atoms dynamically from an array data using splitAtom?

### DISCUSSION THREAD

**Discussion by bashonregardless** [pmndrs/jotai#2255] [Q&A]
I am trying to create atoms from an array using split atom, but I get typescript and `TypeError:  todoAtoms.map is not a function`.

Below is the code that I used : 

In file *appStateAtom.ts*, I have a declared atoms

```javascript
// In response from server I get the below todo object. Here, I am using it directly(mocked) for simplicity
export const todo = [
    {
        task: 'help the town',
        done: false,
    },
    {
        task: 'feed the dragon',
        done: false,
    },
]

export const todoStateAtom = atom([] as typeof todo)
export const todoAtomsAtom = atom([])
```

In file *appDispatcher.ts*, I set this atom

```javascript
 const fetchTodoAtom = atom(null, async (_get, set) => {
     try {
         const fetchTodoRes = await fetchTodo() // here I make api call

        // items are first populated from a server request
         set(
             todoStateAtom,
             fetchTodoRes.data
         )
      } catch (error) {
          console.log(error)
     }
  )


 const setupTodoAtom = atom(
        (get) => get(todoTableDataAtom),
        async (_get, set) => {
            try {
                await set(fetchTodoAtom)
                set(todoAtomsAtom, splitAtom(todoStateAtom))  // I get typescript error here
            } catch (error) {
                console.log(error)
            }
        },
    )
```

Then I want to use `todoAtomsAtom` in my component

```javascript
import React, { useCallback, useEffect, useState } from 'react'

import { dispatcherStateAtom } from '@/jotai_model'
import { useAtom, useAtomValue, useSetAtom } from 'jotai'
import { todoAtomsAtom } from '@/jotai_model/atoms/todoState'
import { Todo } from '@/modules/todo/Todo'

export const TodoContainer = () => {
    const { setupTodoAtom } = useAtomValue(dispatcherStateAtom)
    const setupTodo = useSetAtom(setupTodoAtom)

    const [todoAtoms, dispatch] = useAtom(todoAtomsAtom)

    useEffect(() => {
        async function fetchSetupData() {
            await setupTodo()
        }
        fetchSetupData()
    }, [])

    return (
        <div>
            {todoAtoms.map((todoAtom, idx) => (
                <Todo
                    key={idx}
                    todoAtom={todoAtom}
                />
            ))}
        </div>
    )
}

```

Where I do `set(todoAtomsAtom, splitAtom(todoStateAtom))`, I get error 

```javascript
Argument of type 'WritableAtom<PrimitiveAtom<{ task: string; done: boolean; }>[], [SplitAtomAction<{ task: string; done: boolean; }>], void>' is not assignable to parameter of type 'SetStateAction<never[]>'.
  Type 'WritableAtom<PrimitiveAtom<{ task: string; done: boolean; }>[], [SplitAtomAction<{ task: string; done: boolean; }>], void>' is missing the following properties from type 'never[]': length, pop, push, concat, and 29 more.ts(2345)
```

And Inside my component I get error
```javascript
TypeError: todoAtoms.map is not a function
```

**Where am I going wrong?**

**Comment by dai-shi** [comment]
Is it TS only issue? Does it work fine without types?

**Comment by bashonregardless** [comment]
My requirement was to create atoms on demand, as I did not know the length of the items in the list beforehand.

To fix this, I init `todoAtomsAtom` as `const todoAtomsAtom = splitAtom(todoStateAtom)` and deleted `set(todoAtomsAtom, splitAtom(todoStateAtom))`.  The error did not occur after this.

But then I realised that, setting `todoAtomsAtom` in file  _appStateAtom.ts_ and subsequent `useAtom` in `Todo` components would cause all the `Todo` components to use a single state, i.e any action I performed affected all the `Todo` components.

To fix this, I went with the on demand atom creation as given in the docs [here](https://jotai.org/docs/guides/atoms-in-atom).
I was skeptical to take this approach at first, because in the docs its written that "_It might not make any sense_". So, I searched for a while for a better solution.
After implementing this solution with the approach that I have take, I can't help wonder about the benefits of using Jotai in this case. `useState` seems to fit this case as well.

I am not sure if it was a TS only issue. I do have `strict: true` in tsconfig.

**Reply by dai-shi** [reply]
Can you create a small code snippet from your latest version and paste it here? Or https://www.typescriptlang.org/play would work too.
